pennylane/templates/subroutines/all_singles_doubles.py

```diff
@@ -0,0 +1,204 @@
+            )
+
+        if not singles and not doubles:
+            raise ValueError(
+                "'singles' and 'doubles' lists can not be both empty; got singles = {}, doubles = {}".format(
+                    singles, doubles
+                )
+            )
+
+        if doubles is not None:
+            for d_wires in doubles:
+                if len(d_wires) != 4:
+                    raise ValueError(
+                        "Expected entries of 'doubles' to be of size 4; got {} of length {}".format(
+                            d_wires, len(d_wires)
+                        )
+                    )
+
+        if singles is not None:
+            for s_wires in singles:
+                if len(s_wires) != 2:
+                    raise ValueError(
+                        "Expected entries of 'singles' to be of size 2; got {} of length {}".format(
+                            s_wires, len(s_wires)
+                        )
+                    )
+
+        weights_shape = qml.math.shape(weights)
+        exp_shape = self.shape(singles, doubles)
+        if weights_shape != exp_shape:
+            raise ValueError(f"'weights' tensor must be of shape {exp_shape}; got {weights_shape}.")
+
+        # we can extract the numpy representation here since hf_state can never be differentiable
+        self.hf_state = qml.math.toarray(hf_state)
+        self.singles = singles
+        self.doubles = doubles
+
+        if hf_state.dtype != np.dtype("int"):
+            raise ValueError(f"Elements of 'hf_state' must be integers; got {hf_state.dtype}")
+
+        super().__init__(weights, wires=wires, do_queue=do_queue)
+
+    def expand(self):
+
+        weights = self.parameters[0]
+
+        with qml.tape.QuantumTape() as tape:
+
+            BasisState(self.hf_state, wires=self.wires)
+
+            for i, d_wires in enumerate(self.doubles):
+                qml.DoubleExcitation(weights[len(self.singles) + i], wires=d_wires)
+
+            for j, s_wires in enumerate(self.singles):
+                qml.SingleExcitation(weights[j], wires=s_wires)
+
+        return tape
+
+    @staticmethod
+    def shape(singles, doubles):
+        r"""Returns the shape of the tensor containing the circuit parameters.
+
+        Args:
+            singles (Sequence[Sequence]): sequence of lists with the indices of the two qubits
+                the :class:`~.pennylane.SingleExcitation` operations act on
+            doubles (Sequence[Sequence]): sequence of lists with the indices of the four qubits
+                the :class:`~.pennylane.DoubleExcitation` operations act on
+
+        Returns:
+            tuple(int): shape of the tensor containing the circuit parameters
+        """
+        if singles is None:
+            if doubles is not None:
+                shape_ = (len(doubles),)
+        elif doubles is None:
+            shape_ = (len(singles),)
+        else:
+            shape_ = (len(singles) + len(doubles),)
+
+        return shape_
```

Comment:
I _think_ you don't have a case for `singles=None, doubles=None`: It will go into the first if statement and then not execute any code, which means `_shape` is not defined. I know that this is not an expected case, but a user could do it...

Reply:
I had it but I this case is checked above
```python
if not singles and not doubles:
            raise ValueError(
                "'singles' and 'doubles' lists can not be both empty; got singles = {}, doubles = {}".format(
                    singles, doubles
                )
            )
```
One possibility is to move this check inside the shape function?

Reply:
I would do that, because the shape function is not only called by you in the `init` function, but is something a user could do:

```python
params = np.random.random(size=MyTemplate.shape(singles, doubles))

# in the circuit:
MyTemplate(params)

```